provider/provider.go

```diff
@@ -0,0 +1,303 @@
+func (p *dockerNativeProvider) DiffConfig(ctx context.Context, req *rpc.DiffRequest) (*rpc.DiffResponse, error) {
+	return &rpc.DiffResponse{}, nil
+}
+
+// Configure configures the resource provider with "globals" that control its behavior.
+func (p *dockerNativeProvider) Configure(_ context.Context, req *rpc.ConfigureRequest) (*rpc.ConfigureResponse, error) {
+	return &rpc.ConfigureResponse{
+		AcceptSecrets: true,
+	}, nil
+}
+
+// Invoke dynamically executes a built-in function in the provider.
+func (p *dockerNativeProvider) Invoke(_ context.Context, req *rpc.InvokeRequest) (*rpc.InvokeResponse, error) {
+	tok := req.GetTok()
+	return nil, fmt.Errorf("unknown Invoke token '%s'", tok)
+}
+
+// StreamInvoke dynamically executes a built-in function in the provider. The result is streamed
+// back as a series of messages.
+func (p *dockerNativeProvider) StreamInvoke(
+	req *rpc.InvokeRequest, server rpc.ResourceProvider_StreamInvokeServer) error {
+	tok := req.GetTok()
+	return fmt.Errorf("unknown StreamInvoke token '%s'", tok)
+}
+
+// Check validates that the given property bag is valid for a resource of the given type and returns
+// the inputs that should be passed to successive calls to Diff, Create, or Update for this
+// resource. As a rule, the provider inputs returned by a call to Check should preserve the original
+// representation of the properties as present in the program inputs. Though this rule is not
+// required for correctness, violations thereof can negatively impact the end-user experience, as
+// the provider inputs are using for detecting and rendering diffs.
+func (p *dockerNativeProvider) Check(ctx context.Context, req *rpc.CheckRequest) (*rpc.CheckResponse, error) {
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.Create(%s)", p.name, urn)
+	logging.V(9).Infof("%s executing", label)
+
+	return &rpc.CheckResponse{Inputs: req.News, Failures: nil}, nil
+}
+
+// Diff checks what impacts a hypothetical update will have on the resource's properties.
+func (p *dockerNativeProvider) Diff(ctx context.Context, req *rpc.DiffRequest) (*rpc.DiffResponse, error) {
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.Diff(%s)", p.name, urn)
+	logging.V(9).Infof("%s executing", label)
+
+	oldState, err := plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{
+		KeepUnknowns: true,
+		SkipNulls:    true,
+		KeepSecrets:  true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract old inputs from the `__inputs` field of the old state.
+	oldInputs := parseCheckpointObject(oldState)
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{
+		KeepUnknowns: true,
+		SkipNulls:    true,
+		KeepSecrets:  true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	d := oldInputs.Diff(news)
+
+	if d == nil {
+		return &rpc.DiffResponse{
+			Changes: rpc.DiffResponse_DIFF_NONE,
+		}, nil
+	}
+
+	return &rpc.DiffResponse{
+		Changes: rpc.DiffResponse_DIFF_UNKNOWN,
```

Comment:
Perhaps DIFF_SOME?

Reply:
We discussed this and from @viveklak's experience writing other native providers, the `DIFF_UNKNOWN` is a better catch-all for now?